File: frontend/src/types/seat.ts
/** TypeScript types mirroring the IQueue seat assignment Pydantic schemas. */

export type SeatType = "window" | "aisle" | "middle";
export type SeatStatus = "available" | "occupied" | "reserved" | "blocked";
export type SeatSide = "left" | "right";

export interface SeatMapEntry {
  seat_id: string;
  seat_label: string; // e.g. "3B"
  row_number: number;
  col_number: number;
  seat_type: SeatType;
  side: SeatSide;
  is_near_exit: boolean;
  is_accessibility: boolean;
  status: SeatStatus;
  passenger_name?: string;
  group_id?: string;
  language_preference?: string;
  travel_habit?: string;
  lifestyle_interest?: string;
  needs_accessibility?: boolean;
  preferred_seat_type?: string;
  affinity_score?: number;
  boarding_window?: string;
}

export interface SeatMapSummaryResponse {
  bus_id: string;
  seats: SeatMapEntry[];
  total_seats: number;
  occupied_count: number;
  available_count: number;
}

export interface PassengerContext {
  booking_id: string;
  passenger_name: string;
  group_id?: string;
  language_preference?: string;
  travel_habit?: string;
  lifestyle_interest?: string;
  needs_accessibility: boolean;
  preferred_seat_type?: SeatType;
  preferred_side?: SeatSide;
}

export interface SeatAssignmentResult {
  seat_id: string;
  seat_label: string;
  seat_type: SeatType;
  side: SeatSide;
  row_number: number;
  affinity_score: number;
  boarding_window?: string;
}

export interface SeatAssignRequest {
  bus_id: string;
  passenger: PassengerContext;
}

export interface SeatSwapRequest {
  booking_id_a: string;
  booking_id_b: string;
}

export interface SeatSwapResponse {
  status: string;
  seat_a: string;
  seat_b: string;
}

/** UI state per seat cell in the grid. */
export type SeatCellState =
  | "available"
  | "occupied"
  | "selected" // passenger's current selection
  | "auto_assigned" // AI-recommended seat (pre-highlighted)
  | "group_reserved" // held for group member
  | "accessibility" // accessibility-priority seat
  | "blocked";

File: backend/app/schemas/seat.py
"""Seat assignment schemas — Pydantic v2 request/response models."""

from __future__ import annotations

from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class PassengerContext(BaseModel):
    """Input to the assignment engine — passenger preferences and constraints."""

    model_config = ConfigDict(from_attributes=True)

    booking_id: str = Field(..., description="UUID of the booking")
    passenger_name: str = Field(..., min_length=1, max_length=255)
    group_id: Optional[str] = Field(
        None, description="UUID of the travel group (if any)"
    )
    language_preference: Optional[str] = Field(
        None, description="ISO 639-1 code: 'fil', 'en', 'id', 'vi'"
    )
    travel_habit: Optional[str] = Field(
        None, description="business | leisure | student | family"
    )
    lifestyle_interest: Optional[str] = Field(
        None, description="Comma-separated interests for affinity matching"
    )
    needs_accessibility: bool = Field(
        False, description="Whether passenger requires accessible seating"
    )
    preferred_seat_type: Optional[str] = Field(
        None, description="'window' | 'aisle' | None"
    )
    preferred_side: Optional[str] = Field(
        None, description="'left' | 'right' | None"
    )


class SeatAssignmentResult(BaseModel):
    """Returned by SeatAllocator.assign()."""

    seat_id: str
    seat_label: str = Field(..., description="e.g. '3B'")
    seat_type: str = Field(..., description="window | aisle | middle")
    side: str = Field(..., description="left | right")
    row_number: int
    affinity_score: float
    boarding_window: Optional[str] = Field(
        None, description="HH:MM–HH:MM format"
    )


class SeatMapEntry(BaseModel):
    """One seat in the full bus seat map (used by grid + list views)."""

    seat_id: str
    seat_label: str
    row_number: int
    col_number: int
    seat_type: str
    side: str
    is_near_exit: bool
    is_accessibility: bool
    status: str  # available | occupied | reserved | blocked
    passenger_name: Optional[str] = None
    group_id: Optional[str] = None
    language_preference: Optional[str] = None
    travel_habit: Optional[str] = None
    lifestyle_interest: Optional[str] = None
    needs_accessibility: Optional[bool] = None
    preferred_seat_type: Optional[str] = None
    affinity_score: Optional[float] = None
    boarding_window: Optional[str] = None


class SeatAssignRequest(BaseModel):
    """POST /api/v1/seats/assign request body."""

    bus_id: str = Field(..., description="UUID of the bus")
    passenger: PassengerContext


class SeatSwapRequest(BaseModel):
    """PUT /api/v1/seats/swap — operator use only."""

    booking_id_a: str = Field(..., description="UUID of first booking")
    booking_id_b: str = Field(..., description="UUID of second booking")


class SeatMapResponse(BaseModel):
    """Wrapper for the seat map list response."""

    bus_id: str
    seats: list[SeatMapEntry]
    total_seats: int
    occupied_count: int
    available_count: int
